### ISSUE TITLE
Synchronous distributed tensorflow training runs asynchronously

### ISSUE THREAD

**Issue by GD06** [tensorflow/tensorflow#9596]
### System Information:
- **Debian 4.5.5**
- **TF installed from binary (pip3 install tensorflow-gpu==1.0.1 --user)**
- **TF version: v1.0.0-65-g4763edf-dirty 1.0.1**
- **Bazel version: N.A.**
- **CUDA 8.0 cuDNN v5.1**

### Steps to reproduce
1. Make a directory and download the following files into it:
[training.py](https://gist.github.com/GD06/b88d4b20a4587add984323525bbf4be2) [run.sh](https://gist.github.com/GD06/0bdb36ba6b76070b7cd7fadfd30fce6b)
2. Run the command ./run.sh to simply reproduce this issue.

### Detailed descriptions for the bug

Recently, I tried to deploy the synchronous distributed tensorflow training on the cluster. I followed the tutorial and the inception example to write my own program. The [training.py](https://gist.github.com/GD06/b88d4b20a4587add984323525bbf4be2) is from other user's [implementation](http://ischlag.github.io/2016/06/12/async-distributed-tensorflow/), which follows the same API usage as the official example. I modified it to enable it running on a single machine with multiple GPUs by making them communicate through localhost and mapping each worker to see only one GPU.

The [run.sh](https://gist.github.com/GD06/0bdb36ba6b76070b7cd7fadfd30fce6b) launched three processes. One of them is the parameter server and the others are two workers implemented by between-graph replication. I created the training supervisor by tf.train.Supervisor() to manage multiple sessions in the distributed training for the initialization and synchronization. 

I expect these two workers would synchronize each batch and work in the same epoch. However, the worker 0, which is launched prior to the worker 1, completed the whole training set without waiting for the worker 1. After that, the process of the worker 0 finished training process and exited normally while worker 1 behaved like falling into the deadlock and keep near 0% utilization of CPU and GPU for several hours.

Based on my observation, I suspect these two workers didn't communicate and synchronize at all for the data they passed. I report this problem as a bug because I create the optimizer tf.train.SyncReplicasOptimizer as suggested by the official website and the inception example. However, it seems that the synchronization behaviors, if any, are very strange and the program can not exit normally.

### Source code / logs
Two files:
[training.py](https://gist.github.com/GD06/b88d4b20a4587add984323525bbf4be2): This file contains the source code for the parameter server and workers created to use synchronous distributed optimizers (tf.train.SyncReplicasOptimizer). 
[run.sh](https://gist.github.com/GD06/0bdb36ba6b76070b7cd7fadfd30fce6b): This file launched the parameter server and the workers.
Log:
Please produce according to the steps and look at worker_0_log and worker_1_log

**Comment by andydavis1** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by GD06** [comment]
@andydavis1 Thank you for your kind suggestions! Actually, I found the ([similar problem](http://stackoverflow.com/questions/41293576/distributed-tensorflow-good-example-for-synchronous-training-on-cpus)) on the StackOverflow and these are no any useful solutions so far.

Since I don't think we use the optimizer and supervisor in any incorrect way, I put this as an issue. I sincerely hope there is anyone helping us to understand the reason if it's not a bug.

**Comment by wishfpp** [comment]
@GD06 @andydavis1  I am a rookie in developing DL applications with tensorflow and mxnet. Recently there is a  synchronize problem between my TF workers which I suspect is the same problem mentioned in this issue.
I have run the code uploaded by @GD06 in my server with titanX GPUs and the program also gets blocked while I still can't find the error in your code (just like ours) by myself.
So have you solved the problem? Would you please let us know if there is any update? Thanks very much!

**Comment by GD06** [comment]
@andydavis1 To convince the workers executing asynchronously, I improve the [training.py](https://gist.github.com/GD06/b56d62b718bbd0b033703320e4b214dc) and make worker 1 exist after the first time output information. As expected, the worker 0 continue to execute without waiting to the worker 1.

**Comment by jmchen-g** [comment]
There is a known issue with sync_rep_opt_v2 with the device setter/chooser related to the allocation of the local variable created in the sync_rep_opt_v2. You can either use the old sync_rep_opt or wait until that is fixed (where we might need to have similar signature anyway). Thanks.

**Comment by GD06** [comment]
@jmchen-g Thanks a lot! Would you mind reopening this issue? I think this issue should be clarified that it's not caused by incorrect API uses. Otherwise, other developers will be still very confused in such examples as in the official guide and tensorflow examples.

**Comment by GD06** [comment]
@andydavis1 Would you mind reopening this issue? Or at least you need to label this issue as the same/similar to the known issue pointed by @jmchen-g.
Thanks!

**Comment by HokyungLee** [comment]
Can I have a question related to thread
I have no idea about this page a little, so Is there multi-thread code in this page?
Please, I really wanna solid information that I catch up synchronous & threading programming
Thanks.

**Comment by GD06** [comment]
@HokyungLee 
There is no any multi-thread code in this page. The script run.sh is only used to start the distributed training by starting the worker and parameter server in a multi-process way.
We regard each GPU as a worker and a CPU process as a parameter server.

**Comment by HokyungLee** [comment]
Thank you for answering
It makes me help a lot @GD06

**Comment by ahfant** [comment]
It seems that this issue is still closed. Is it being tracked under a separate ID or has it been resolved?
